### PR TITLE
fix(app): add logic for component inheritance

### PR DESCRIPTION
**Pull request by tripodsan**
Issue:
relates to #45

## What I did
- added logic to component loader to resolve the script based on `sling:resourceSuperType`
- import some of the aem-core-components in the aem-kitchen-sink and created extended component (AEM Text).
- improved webpack component loader to better detect the resource type
-moved some of the aem-core-components below `jcr_root` so that the resource type can be detected

## How to test
- build and run the aem-kitchensink-example
- there should be a functional `AEM Text` story

## Still TODO:
- improve the way to determine the resource type
- find a way to add multiple template roots to HTL engine

### CODE REVIEW

File: app/aem/src/client/preview/helpers/ComponentLoader.ts [hunk 1]
```diff
@@ -1,6 +1,37 @@
...
+  }
+
+  /**
+   * Resolves the HTL script for the given resource type, respecting the `sling:resourceSuperType`
+   * property.
+   *
+   * @param {string} type Resource Type.
+   * @return {function} the script function or {@code null}
+   */
+  resolveScript(type) {
```
Comment (CodeByAlex) [inline app/aem/src/client/preview/helpers/ComponentLoader.ts:27]: Could you add a type

Comment (CodeByAlex) [reply app/aem/src/client/preview/helpers/ComponentLoader.ts:27]: Actually add types to all your changes- we should get in the habit of doing that

Comment (CodeByAlex) [reply app/aem/src/client/preview/helpers/ComponentLoader.ts:27]: Please :)

File: app/aem/src/server/aem-component-loader.js [hunk 2]
```diff
@@ -31,20 +34,52 @@ const getRequiredClientLibs = componentDir => {
...
-  const resourceType = this.context.substring(this.rootContext.length + 1);
-  const json = JSON.parse(toJson(source));
-  const pathBaseName = basename(this.context);
+/**
+ * Computes the resource type given the webpack context and root-context.
+ * This is based on heuristic and best practices. it would be better if
+ * the respective roots could be specified via configuration.
+ *
+ * @param {string} rootContext compilation context (i.e. the directory of the compilation)
+ * @param {string} context compile context (i.e. the directory of the .content.xml)
```
Comment (CodeByAlex) [inline app/aem/src/server/aem-component-loader.js:43]: Great comments!

File: app/aem/src/server/aem-component-loader.js [hunk 3]
```diff
@@ -31,20 +34,52 @@ const getRequiredClientLibs = componentDir => {
...
+  if (idx >= 0) {
+    // remove all segments including node_modules
+    segs.splice(0, idx + 1);
+  }
+  // if component is below a jcr_root
+  idx = segs.indexOf('jcr_root');
+  if (idx >= 0) {
+    // remove all segments including jcr_root
+    segs.splice(0, idx + 1);
+    if (segs[0] === 'apps' || segs[1] === 'libs') {
```
Comment (CodeByAlex) [inline app/aem/src/server/aem-component-loader.js:59]: Could you make these strings constants?

File: app/aem/src/client/preview/helpers/ComponentLoader.ts (hunk 1 above)
Comment (tripodsan) [reply app/aem/src/client/preview/helpers/ComponentLoader.ts:27]: +1 

sure. I completely forgot. I usually never code in typescript.

**Comment by tripodsan** [comment]
@CodeByAlex what is the problem with the failing tests ?

**Comment by CodeByAlex** [comment]
> LGTM

It does look like the build is failing though - could you check that it builds locally?

**Comment by CodeByAlex** [comment]
@tripodsan it’s netlify checking that the build passed

**Pull request merged**